src/compilation/compiler.rs

```diff
@@ -310,22 +310,49 @@ impl Compiler {
         match statement {
             Statement::VarDeclarations(v) => self.compile_var_declarations(v),
             Statement::WhileLoop(w) => self.compile_while_loop(w),
-            Statement::Expr(e) => self.compile_expr_statement(e),
+            Statement::Assignment(a) => self.compile_assignment_statement(a),
             Statement::Return(r) => self.compile_ret_statement(r),
         }
     }
 
     /// Converts Flick's [Type] enum to llvm-sys's [LLVMTypeRef].
     unsafe fn to_llvm_type(&self, t: Type) -> LLVMTypeRef {
         match t {
-            Type::I64 => LLVMIntTypeInContext(self.context, 64),
+            Type::Int { width } => LLVMIntTypeInContext(self.context, width),
             Type::Void => LLVMVoidTypeInContext(self.context),
         }
     }
 
     /// Compiles a while loop, assuming the LLVM builder is building inside a function body.
-    fn compile_while_loop(&self, _while_loop: &WhileLoop) {
-        todo!()
+    unsafe fn compile_while_loop(&mut self, while_loop: &WhileLoop) {
+        let cur_func = self.get_cur_function();
+        let cond_block = LLVMAppendBasicBlockInContext(self.context, cur_func, cstr!("cond"));
+        let loop_block = LLVMAppendBasicBlockInContext(self.context, cur_func, cstr!("loop"));
+        // Build block to go to after loop is done executing
+        let after_block = LLVMAppendBasicBlockInContext(self.context, cur_func, cstr!("after"));
+
+        LLVMBuildBr(self.builder, cond_block);
+        // todo: consider deleting?
+        // Insert an explicit fall through from the current block to the loop_block.
+        LLVMPositionBuilderAtEnd(self.builder, cond_block);
+
+        // TODO: types: Should have type bool
+        // Build a boolean LLVMValueRef with value 'condition != 0', which represents whether we should continue looping
+        let condition = self.compile_expr(&while_loop.condition, Type::Int { width: 1 });
+        LLVMBuildCondBr(self.builder, condition, loop_block, after_block);
+
+        // Start insertion in loop_block.
+        LLVMPositionBuilderAtEnd(self.builder, loop_block);
+
+        self.scope_manager.enter_scope();
+        for statement in while_loop.body.iter() {
+            self.compile_statement(statement);
+        }
+        self.scope_manager.exit_scope();
```

Comment:
Just a thought for the future: we could extract this into a `compile_body()` function bc it's repeated whenever we compile a body. No need to do it now though

Reply:
Oh actually I like that let me add that as a commit

Reply:
Actually nvm

Reply:
Actually let's implement this later I think. We only use this exact code once right now, but once we add if statements, for loops, etc it definitely make sense to split this into a compile_body function. Right now we have this code in compile_while_loop, and we have similar but different (we set up function arguments after entering the scope but before compiling the statements) code in compile_func_def.